src/engine/TPA.cc

```diff
@@ -1403,74 +1425,77 @@ VerificationResult TransitionSystemNetworkManager::solve() && {
     initNetwork();
 
     auto current = getChainStart();
+    activePath.push(getIncomingEdge(current, 0));
     while (true) {
+        if(getNode(current).blocked_children == getNode(current).children.size()){
+            getNode(current).blocked_children = 0;
+            updateRestrictions(current);
+        }
         auto [res,explanation] = queryTransitionSystem(networkMap.at(current));
         if (reachable(res)) {
             if (current == getChainEnd()) {
                 return {VerificationAnswer::UNSAFE, computeInvalidityWitness()};
             }
-            EId nextEdge = getOutgoingEdge(current);
-//            std::cout << "Reached states in node " << current.id << ": " << logic.pp(explanation) << std::endl;
-            getNode(current).trulyReached = explanation;
-            PTRef nextConditions = getInitialStates(graph.getTarget(nextEdge));
-            auto [edgeRes, edgeExplanation] = queryEdge(nextEdge, explanation, nextConditions);
-            if (reachable(edgeRes)) { // Edge propagates forward
-                auto next = graph.getTarget(nextEdge);
-                networkMap.at(next).solver->resetInitialStates(edgeExplanation);
-                current = next;
-//                std::cout << "Propagating forward query to " << next.id << ": " << logic.pp(edgeExplanation) << std::endl;
-                continue; // Information has been propagated to the next node, continue querying that node
-
-            } else { // Edge cannot propagate forward
-                getNode(current).trulyReached = PTRef_Undef;
-//                std::cout << "Edge cannot propagate forward from: " << logic.pp(edgeExplanation) << std::endl;
-                getNode(current).solver->updateQueryStates(logic.mkNot(edgeExplanation));
-                continue; // Repeat the query for the same TS with stronger query
+            while(getNode(current).blocked_children < getNode(current).children.size()){
+                auto nextEdge = (*getOutgoingEdges(current))[getNode(current).blocked_children];
+                getNode(current).trulyReached = explanation;
+                PTRef nextConditions = getInitialStates(graph.getTarget(nextEdge));
+                auto [edgeRes, edgeExplanation] = queryEdge(nextEdge, explanation, nextConditions);
+                getNode(current).blocked_children ++;
+                if (reachable(edgeRes)) { // Edge propagates forward
+                    auto next = graph.getTarget(nextEdge);
+                    networkMap.at(next).solver->resetInitialStates(edgeExplanation);
+                    activePath.push(nextEdge);
+                    current = next;
+                    break; // Information has been propagated to the next node, switch to the new node
+
+                } else { // Edge cannot propagate forward
+                    getNode(current).trulyReached = PTRef_Undef;
```

Comment:
This seems premature.
Truly reached states should be reset only after *all* outgoing edges has been blocked for them, not after one edge has been blocked.

Reply:
Moved to the state when everything is blcoked